jcl/src/openj9.sharedclasses/share/classes/com/ibm/oti/shared/SharedClassURLHelperImpl.java

```diff
@@ -103,7 +103,7 @@ public boolean setMinimizeUpdateChecks() {
 	}
 
 	@Override
-	public synchronized byte[] findSharedClass(String partition, URL path, String className) {
+	public byte[] findSharedClass(String partition, URL path, String className) {
```

Comment:
newJarFileCheck() needs internal synchronization. Also the SharedClassFilter doesn't mention synchronization, but removing `synchronized` changes the behavior for calling acceptFind() and acceptStore(), these calls should remain protected.

Reply:
> Also the SharedClassFilter doesn't mention synchronization, but removing synchronized changes the behavior for calling acceptFind() and acceptStore(), these calls should remain protected.

You mean removing `synchronized` from getSharingFilter()/setSharingFilter(), then protect the calls of getSharingFilter(), setSharingFilter(), acceptFind() and acceptStore() within readWriteLock ?

Reply:
> newJarFileCheck() needs internal synchronization. 

I am thinking about changing `jarFileNameCache` from a regular `Set` to `ConcurrentHashMap.newKeySet()`.

Reply:
It doesn't need to use the readWriteLock(), but something should keep the synchronized behavior of calling acceptFind() and acceptStore() the same. Similarly, getSharingFilter()/setSharingFilter() are fine the way they are since they are already synchronized.

Reply:
> I am thinking about changing jarFileNameCache from a regular Set to ConcurrentHashMap.newKeySet()

As long as it doesn't cause a measurable increase in footprint.

Reply:
Changed `jarFileNameCache` to a concurrent Set and wrap acceptFind()/acceptStore() inside a synchronized block.